sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala

```diff
@@ -288,29 +285,105 @@ class HadoopTableReader(
     }
   }
 
+  /**
+   * The entry of creating a RDD.
+   */
+  private def getRDD(
+    inputClassName: String,
+    localTableDesc: TableDesc,
+    inputPathStr: String): RDD[Writable] = {
+    if (isCreateNewHadoopRDD(inputClassName)) {
+      createNewHadoopRdd(
+        localTableDesc,
+        inputPathStr,
+        inputClassName)
+    } else {
+      createHadoopRdd(
+        localTableDesc,
+        inputPathStr,
+        inputClassName)
+    }
+  }
+
   /**
    * Creates a HadoopRDD based on the broadcasted HiveConf and other job properties that will be
    * applied locally on each slave.
    */
   private def createHadoopRdd(
     tableDesc: TableDesc,
     path: String,
-    inputFormatClass: Class[InputFormat[Writable, Writable]]): RDD[Writable] = {
+    inputClassName: String): RDD[Writable] = {
 
     val initializeJobConfFunc = HadoopTableReader.initializeLocalJobConfFunc(path, tableDesc) _
 
     val rdd = new HadoopRDD(
       sparkSession.sparkContext,
       _broadcastedHadoopConf.asInstanceOf[Broadcast[SerializableConfiguration]],
       Some(initializeJobConfFunc),
-      inputFormatClass,
+      getInputFormat(inputClassName),
       classOf[Writable],
       classOf[Writable],
       _minSplitsPerRDD)
 
     // Only take the value (skip the key) because Hive works only with values.
     rdd.map(_._2)
   }
+
+  /**
+   * Creates a HadoopRDD based on the broadcasted HiveConf and other job properties that will be
+   * applied locally on each slave.
+   */
+  private def createNewHadoopRdd(
+    tableDesc: TableDesc,
+    path: String,
+    inputClassName: String): RDD[Writable] = {
+
+    val initializeJobConfFunc = HadoopTableReader.initializeLocalJobConfFunc(path, tableDesc) _
```

Comment:
Looks we can just directly call rather then making a function and call it later.